pilot/proxy/envoy/discovery.go

```diff
@@ -27,12 +27,78 @@ import (
 
 	restful "github.com/emicklei/go-restful"
 	"github.com/golang/glog"
-	multierror "github.com/hashicorp/go-multierror"
-
+	"github.com/hashicorp/go-multierror"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 	"istio.io/istio/pilot/model"
 	"istio.io/istio/pilot/proxy"
 )
 
+const (
+	metricsPath          = "/metrics"
+	metricsNamespace     = "pilot"
+	metricsSubsystem     = "discovery"
+	metricLabelCacheName = "cache_name"
+	metricLabelMethod    = "method"
+)
+
+var (
+	cacheSizeGauge = prometheus.NewGaugeVec(
```

Comment:
[non-blocking] just thinking out loud: will we ever want to track metrics per Pilot release?  should we add version info as a label?

Reply:
that would actually be good.. or atleast SHA? So that we can see what difference things have made?

Reply:
done.

Reply:
the approach taken here is fine, though slightly different than i was anticipating. it looks like you added a new /version endpoint. that's probably very useful too.

I was suggesting altering each metric to include a version dimension, via something like:
```
cacheHitCounter = prometheus.NewCounterVec(
 	prometheus.CounterOpts{
 		Namespace: metricsNamespace,
 		Subsystem: metricsSubsystem,
 		Name:      "cache_hit",
 		Help:      "Count of cache hits for a particular cache within Pilot",
 	}, []string{metricLabelCacheName, **pilotVersionLabel**})
```

Reply:
@douglas-reid yup, I understood that when reading your comment and then quickly forgot it when I wrote the code :).  Will create a follow-on to fix.

Reply:
@douglas-reid created #2167